### PR TITLE
 bloom filter

### PR DESCRIPTION
**Pull request by gluk256**
initial version

### CODE REVIEW

**Comment by gluk256** [comment]
It seems, the errors are not related to my changes.

File: libwhisper/BloomFilter.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+void TopicBloomFilter::addRaw(AbridgedTopic const& _h)
+{
+	*this |= _h;
+	for (unsigned i = 0; i < CounterSize; ++i)
+		if (isBitSet(_h, i))
+		{
+			if (m_refCounter[i] != numeric_limits<uint16_t>::max())
+				m_refCounter[i]++;
+			//else: overflow
```
Comment (subtly) [inline libwhisper/BloomFilter.cpp]: Could an exception be thrown?

**Comment by gluk256** [comment]
I don't think it is a severe error. First, it is very unlikely, that somebody will watch more than 65536 topics. It will only happen if somebody only adds topics to the filter, but does not remove them. It does not make any sense, though -- several hundreds of them will already fill the entire bloom filter with very high probability. 
But even in case of overflow, it will cause no immediate problems.The error will only arise after at least 65536 of them will be removed later. Since overflow only can happen if topics are never removed, even if overflow happens, it will probably not cause any problem in the future. That's why I decided to leave it like this.

**Comment by chriseth** [comment]
Is this an attack vector a malicious DApp could use?

**Comment by gluk256** [comment]
At the moment it should not be a problem.

**Comment by gavofyork** [comment]
the only thing we can sign up to protecting users from regarding malicious dapps is the use of secrets. everything else is fair game for now.

File: libwhisper/BloomFilter.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+void TopicBloomFilter::removeRaw(AbridgedTopic const& _h)
+{
+	for (unsigned i = 0; i < CounterSize; ++i)
+		if (isBitSet(_h, i))
+		{
+			if (m_refCounter[i])
+				m_refCounter[i]--;
+
+			if (!m_refCounter[i])
+				(*this)[i/8] &= ~c_mask[i%8];
```
Comment (gavofyork) [inline libwhisper/BloomFilter.cpp]: whitespace. (`i / 8`, `i % 8`).

File: libwhisper/Common.cpp [hunk 3]
```diff
@@ -84,6 +85,26 @@ bool TopicFilter::matches(Envelope const& _e) const
 	return false;
 }
 
+TopicFilter::TopicFilter(RLP const& _r)
+{
+	for (RLP i: _r)
```
Comment (gavofyork) [inline libwhisper/Common.cpp]: `RLP const&`

File: libwhisper/Common.cpp [hunk 4]
```diff
@@ -84,6 +85,26 @@ bool TopicFilter::matches(Envelope const& _e) const
...
+	{
+		m_topicMasks.push_back(TopicMask());
+		for (RLP j: i)
```
Comment (gavofyork) [inline libwhisper/Common.cpp]: `RLP const&`

File: libwhisper/Common.cpp [hunk 5]
```diff
@@ -84,6 +85,26 @@ bool TopicFilter::matches(Envelope const& _e) const
...
+			m_topicMasks.back().push_back(j.toPair<FixedHash<4>, FixedHash<4>>());
+	}
+}
+
+AbridgedTopic TopicFilter::exportBloom() const
+{
+	AbridgedTopic ret;
+	for (TopicMask const& t: m_topicMasks)
+		for (auto i: t)
```
Comment (gavofyork) [inline libwhisper/Common.cpp]: `auto const&`

**Comment by gavofyork** [comment]
whitespace and minor improvements to existing code.

**Pull request merged**